src/vm/scanning.rs

```diff
@@ -98,14 +98,14 @@ pub trait ObjectTracerContext<VM: VMBinding>: Clone + Send + 'static {
 ///     it needs `Send` to be sent between threads.  `'static` means it must not have
 ///     references to variables with limited lifetime (such as local variables), because
 ///     it needs to be moved between threads.
-pub trait RootsWorkFactory<ES: Edge>: Clone + Send + 'static {
-    /// Create work packets to handle root edges.
+pub trait RootsWorkFactory<SL: Slot>: Clone + Send + 'static {
+    /// Create work packets to handle root slots.
     ///
-    /// The work packet may update the edges.
+    /// The work packet may update the slots.
     ///
     /// Arguments:
-    /// * `edges`: A vector of edges.
-    fn create_process_edge_roots_work(&mut self, edges: Vec<ES>);
+    /// * `slots`: A vector of slots.
+    fn create_process_slot_roots_work(&mut self, slots: Vec<SL>);
```

Comment:
That makes sense.  "slot roots" is not the opposite of "pinning roots".

-   The opposite of "slot roots" (or root slots) is "node roots" (or root nodes), and we already have `ProcessRootNodes`.
-   The opposite of "pinning roots" is "non-pinning roots".

There is one missing piece of the puzzle:  Root nodes that can actually be moved.  The current `RootsWorkFactory` does not allow that, but I already anticipated that (See https://github.com/mmtk/mmtk-core/issues/710), and the Android binding (which @k-sareen is developing) needs that to scan and update stack roots and global roots.  But the problem is, such roots cannot be given as a list of `ObjectReference` or a list of `Slot`.

I think we should keep the word "slots" in the function name so that we can distinguish it from root nodes.  `create_process_root_slots_work` sounds good to me.  I'll use that name in this PR.  In the future, however, we may refactor the `RootsWorkFactory` again to address https://github.com/mmtk/mmtk-core/issues/710.

Reply:
The other functions like `create_process_(t)pinning_roots_work` focus on the semantics of pinning/tpining, rather than how the roots are passed as arguments. It still feels strange as `create_process_root_slots_work` cares about the arguments rather than the semantics.

Reply:
It's a bit subtle.  The fact that a root is represented as `Slot` implies it can be updated (using `slot.store`.  i.e. It's not pinned), and the fact that a root is represented as an `ObjectReference` implies it cannot be updated (i.e. pinned).  `create_process_roots_work` is just too general because the word "roots" covers both pinning roots and non-pinning roots.

One way to reconcile is renaming the functions for nodes, too:

|old|new|
|---|---|
|`create_process_edge_roots_work`|`root_slots`|
|`create_process_pinning_roots_work`|`pinning_root_nodes`|
|`create_process_tpinning_roots_work`|`tpinning_root_nodes`|

(Lets' just omit the repeating `create_process_` and `_work`.)

In this way, we mention both the form (slots / nodes) and the semantics (unrestricted / pinning / tpinning). What do you think about this?

Reply:
We should minimise changes to the existing interface. The only thing that needs to be fixed here is 'edge roots', which is a clear misnomer. 

If we introduced the API in this PR, what you proposed were okay. But for existing APIs, the changes seem unnecessary.

Reply:
> We should minimise changes to the existing interface. The only thing that needs to be fixed here is 'edge roots', which is a clear misnomer.
> 
> If we introduced the API in this PR, what you proposed would be okay. But for existing APIs, the changes seem unnecessary.

Then I think we can just go with `create_process_slot_roots_work` this time.  It's not ideal, but the VM binding developers can simply use search/replace to change `edge` to `slot` which will change `create_process_edge_roots_work` to `create_process_slot_roots_work`.

Reply:
> Then I think we can just go with create_process_slot_roots_work this time. 

'slot root' is a new misnomer as well. If we change from one misnomer to another misnomer, we could just not do the change at all. In that case, at least we minimise changes to the existing interface.

Reply:
Then I'll change it to `create_process_root_slots_work`.

Reply:
Then it is back to this question.
> The other functions like create_process_(t)pinning_roots_work focus on the semantics of pinning/tpining, rather than how the roots are passed as arguments. It still feels strange as create_process_root_slots_work cares about the arguments rather than the semantics.

Reply:
Then we have two solutions.  One is renaming those functions as well as I described, which resulted in the table in [this comment](https://github.com/mmtk/mmtk-core/pull/1134#discussion_r1606219835).  The other is leaving them unchanged for now.  That's not ideal, but we don't introduce new misnomers.

Reply:
I am okay with the latter -- just leave it unchanged before we can have a good fix for it.